01-resources-v2/proposal.md

```diff
@@ -0,0 +1,285 @@
+  vs. static usage patterns, and will get more powerful with #684.
+
+
+## Changes to Versioned Artifact resources
+
+* Add a `spaces` actopm, which is used to discover spaces.
+
+  It's not a verb, partly because most verbs would be ambiguous with `check`,
+  and partly because the API is practically stateless; it just returns whatever
+  spaces there are.
+
+* Add a `destroy` action, which looks like `put` (can be given files and params
+  to determine what to destroy) but returns the set of versions that it
+  deleted.
+
+  This is to support bulk deletes, e.g. to garbage collect intermediate
+  artifacts after a final build is shipped.
+
+* Change `check` and `get` to always run against a particular space, given by
+  the request payload.
+
+* Change `check` to include metadata for each version. Change `get` and `put` to
+  no longer return it.
+
+  This way metadata is always immediately available, and only comes from one
+  place.
+
+  The original thought was that metadata collection may be expensive, but so far
+  we haven't seen that to be the case.
+
+* Change `get` script to no longer return a version, since it's always given one
+  now. As a result, `get` no longer has a response; it just succeeds or fails.
+
+* Change `get` and `put` to run with the bits as their working directory, rather
+  than taking the path as an argument. This was something people would trip up
+  on when implementing a resource.
+
+* Change `put` to emit a list of versions, rather than just one.
+
+  TODO: it's undecided whether `put` should be given a single space, return a
+  single space, or return many spaces + versions.
+
+  Technically the `git` resource may push many commits, so this is necessary to
+  track them all as outputs of a build. This could also support batch creation.
+
+  To ensure `check` is the source of truth for ordering, the versions returned
+  by `put` are not order dependent. A `check` will be performed to discover
+  them in the correct order, and then each version will be saved as an output
+  of the build. The latest version of the set will then be fetched.
+
+* Change `put` to write its JSON response to a specified file, rather than
+  `stdout`, so that we don't have to be attached to process its response.
+
+  This is one of the few ways a build can error after the ATC reattaches
+  (`unexpected end of JSON`). With it written to a file, we can just try to read
+  the file when we re-attach after seeing that the process exited. This also
+  frees up stdout/stderr for normal logging, which has been an occasional
+  pitfall during resource development/debugging.
+
+
+## Introduction of "notifier" resource interface
+
+A resource may now implement a "notifier" interface. Notifications must be
+explicitly enabled in the pipeline. This is so that a resource type can't
+suddenly decide to include a notifier and start doing surprising things.
+
+A notifier has a single hook, `notify`, and receives some sort of JSON payload
```

Comment:
is this not already solved by put-only resources/artifacts? does it need to be a first class citizen?

Reply:
I wouldn't say it's "solved", from a design perspective. It's *doable*, but there are many drawbacks:

* Resources are meant to provide meaningful abstractions. If a whole class of them circumvent the majority of the interface, and act as glorified tasks instead, there's not much meaning to their use in a pipeline - they no longer describe a dependency lifecycle, they describe third-party hooks to execute in artbitrary places.
* "Partial" resource implementations makes it impossible to know what to expect from third-party resources. I don't think it's great to be limited in what verbs you can use, either because the author didn't have the time/motivation to implement the full thing or because it doesn't fit the interface well.
* Using an entire resource just for notifications results in a ton of clutter, both in the config and in the UI.

Reply:
I see where you're coming from. I started rebutting by writing this out as interface composition but when I did the difference between the two was immediately apparent: a notifier doesn't emit versions.

Reply:
Yeah sorry, the Elm mainly helps me right now - it's easier to move around which types goes into which function. Once things gel a bit the doc will get more detailed/accurate. :)

Reply:
Couple of questions on this:
- Notifiers will still be visualised in the pipeline so that it's apparent if it has failed?
- Are notifiers only fed a JSON describing just the events they are "subscribed" to? In your example (`01-resources-v2/notifications.yml`), it seems like the slack notification would trigger and send a message if either the whole pipeline succeeded, or any part of it failed?

If I've understood things correctly in my last question, and you can't use notifier resources in `on_success`/`on_failure` as we do today, I have a few pipelines in production now that would not be able to replace the current slack notification resource with the new "notifier" version. Because, the pipelines aggregate information from multiple "get" steps, piece it together in one task, and output it to slack as the last "put" step.

Reply:
Thanks for the information! Personally Regarding a new `notify:` step type, I think it's a good idea to try to keep the number of keywords to a minimum (as far as possible). But perhaps what I'd suggest is that we skip this:

```
A notifier has a single hook, `notify`[...]
```

What if notify is just another interface that can be implemented by resources, which then makes them eligible for the `notify:` configuration (as you have drafted in your example). But if it also implements `put:`, you can continue to use it as is currently done in pipelines (with custom/aggregated text)?

Perhaps I don't have the correct vision, but impression from the resources I've used (and some I've written) is that only implementing e.g. `check` and `get`, but not `put`, and the other way around, are valid use cases and "working as intended". If `check` and/or `get` are no-op, then I think it would make sense to drop the requirement that resources return versions also. 

Since resources v2 will return `info` with all the interfaces they support, my thinking was that the list of interfaces implemented by the resource could determine how it's represented in Concourse. So e.g. the `slack-notification-resource` could implement only `info, `notify` and `put`, and therefor not return any version, only references to builds and steps that triggered a `notify`, and the metadata emitted at the time?

Reply:
@itsdalmo I prefer to not overload verbs when possible, especially if they would have a different "function signature" (e.g. `put` for notifier doesn't return versions). I think today's pattern of partial resource implementations has weakened the strength of the abstraction and achieved a local maxima. It makes it harder to reason about resources as meaningful objects - they become just bags of verbs.

So I'd rather take this opportunity to introduce more meaningful concepts (i.e. notifications/status reports) rather than overload the verbs already being used for versioned artifacts. Having more verbs which always mean one thing, and are always tied to a meaningful interface, will be cleaner in the long run, even though mechanically they may be fairly similar.

With the addition of `/info` you can then have a single image that supports both the notifier interface and the versioned artifact interface. This would be useful for things like pull request resources, which would use the notifier interface to reflect the build status back to GitHub, while using the versioned artifact interface for the actual brunt of the work (finding PRs/commits and fetching them).